File: pytw/demos/camera/Perspective.cc
/* Contrast of perspective as a function of FOVY. Both scenes are viewing
   a wire cube of the same size.  The camera moved farther from the scene
   and the NEAR is changed so that the image plane is the same distance
   from the cube in each scene.  The aspect ratio of the cameras is the
   same.  The only difference in the camera shapes is the FOVY.

Implemented Fall 2003
Scott D. Anderson
*/

#include <stdio.h>
#include <math.h>
#include <tw.h>

#define GAP  25      /* gap between subwindows */
int leftWinWidth=300;
int leftWinHeight=300;
int rightWinWidth=300;
int rightWinHeight=300;

void parentDisplay(void) {
    twDisplayInit();
    glFlush();
    glutSwapBuffers();
}

GLfloat aspectRatio(int w, int h) {
    return ((GLfloat) w) / ((GLfloat) h);
}

void leftDisplay(void) {
    twDisplayInit(0.8,1.0,1.0);
    glMatrixMode(GL_PROJECTION);
    glLoadIdentity();
    gluPerspective(90,aspectRatio(leftWinWidth,leftWinHeight),1,3);
    glMatrixMode(GL_MODELVIEW);
    glLoadIdentity();
    gluLookAt(0,0,2, 0,0,0, 0,1,0);

    twColorName(TW_RED);
    glutWireCube(1);
  
    glFlush();
    glutSwapBuffers();
}

void rightDisplay(void) {
    twDisplayInit(0.8,1.0,0.8);
    glMatrixMode(GL_PROJECTION);
    glLoadIdentity();
    const GLfloat M = 3;        // multiplication factor
    // The following code calculates a fovy such that the focal point is M 
    // times farther from the image plane than in the leftDisplay.  The
    // calculation is based on right triangle with legs of length 1 and M,
    // the fovy is double that angle, converted to degrees.
    GLfloat fovy = 2*atan(1/M)*180/M_PI;
    printf("fovy = %f degrees\n",fovy);
    gluPerspective(fovy,aspectRatio(rightWinWidth,rightWinHeight),M,2+M);
    glMatrixMode(GL_MODELVIEW);
    glLoadIdentity();
    gluLookAt(0,0,1+M, 0,0,0, 0,1,0);

    twColorName(TW_MAGENTA);
    glutWireCube(1);
  
    glFlush();
    glutSwapBuffers();
}

int parentWin, leftWin, rightWin;

void redisplayAll(void) {
    glutSetWindow(leftWin);
    glutPostRedisplay();
    glutSetWindow(rightWin);
    glutPostRedisplay();
}

int main(int argc, char** argv) {
    glutInit(&argc, argv);
    glutInitDisplayMode(GLUT_RGB | GLUT_DEPTH | GLUT_DOUBLE);
    glutInitWindowSize(leftWinWidth+rightWinWidth+GAP*3, 
                       leftWinHeight+GAP*2); //parent window
    parentWin = glutCreateWindow(argv[0]);
    glutDisplayFunc(parentDisplay);
    twBoundingBox(-1,1,-1,1,-1,1); // keeps twMainInit quiet
    twMainInit();

    /*glutCreateSubWindow returns the ID of the sub-window and takes 
      arguments:(int parent,int x,int y,int width,int height).
      Origin at x,y; height equals distance from y _down_ */
    leftWin = glutCreateSubWindow(parentWin,GAP,GAP,leftWinWidth,leftWinHeight);

    /*Each time you create a new window you need to give it its own 
      callbacks and initialization. These can be the same as those
      from other windows but you must call them each time*/
    twBoundingBox(-10,10,-10,10,-10,10);
    glutDisplayFunc(leftDisplay);
    twMainInit(); 

    rightWin = glutCreateSubWindow(parentWin,GAP*2+leftWinWidth,GAP,rightWinWidth,rightWinHeight);
    glutDisplayFunc(rightDisplay);
    twMainInit();
    glLineWidth(2);
  
    redisplayAll(); 
    glutMainLoop();
    return 0;
}

File: pytw/demos/camera/Perspective.py
'''Contrast of perspective as a function of FOVY. Both scenes are viewing
   a wire cube of the same size.  The camera moved farther from the scene
   and the NEAR is changed so that the image plane is the same distance
   from the cube in each scene.  The aspect ratio of the cameras is the
   same.  The only difference in the camera shapes is the FOVY.

Implemented Fall 2003
Scott D. Anderson

Ported to Python, Fall 2009
'''

import sys
import math                     # for atan and others

from TW import *

## ================================================================

GAP = 25      			# gap around subwindows
leftWinWidth=300;
leftWinHeight=300;
rightWinWidth=300;
rightWinHeight=300;

def parentDisplay():
    twDisplayInit();
    glFlush();
    glutSwapBuffers();

def aspectRatio(w, h):
    '''Returns the aspect ratio of a rectangle, given width w and height h.'''
    return float(w)/float(h)

LeftWinEyeZ = 2

def leftDisplay():
    '''display function for the left subwindow'''
    twDisplayInit(0.8,1.0,1.0);
    ar = aspectRatio(leftWinWidth,leftWinHeight)
    eyedist = LeftWinEyeZ - 1   # from top of frustum
    fovy = 2*math.atan(1/eyedist)*180/M_PI
    if fovy != 90.0:
        print "fovy in left window is not 90!!"
    glMatrixMode(GL_PROJECTION);
    glLoadIdentity();
    gluPerspective(fovy,ar,eyedist,eyedist+2);
    glMatrixMode(GL_MODELVIEW);
    glLoadIdentity();
    ## Eye is 2 units from the origin, so 1.5 units from the near face
    ## of the cube and 2.5 units from the far face, so we see a big
    ## difference in apparent dimensions (the front face is 3/5s the
    ## size of the back face)
    gluLookAt(0,0,LeftWinEyeZ, 0,0,0, 0,1,0);

    twColorName(TW_RED);
    glutWireCube(1);            # from z=0.5 to z=-0.5, etc.
  
    glFlush();
    glutSwapBuffers();

AdditionalEyeDistance = 2.0     # additive term.  Eye in Right window
                                # is this many units farther.

def rightDisplay():
    '''display function for the right subwindow'''
    twDisplayInit(0.8,1.0,0.8);
    ar = aspectRatio(leftWinWidth,leftWinHeight)
    RightWinEyeZ = LeftWinEyeZ + AdditionalEyeDistance
    eyedist = RightWinEyeZ - 1  # from top of frustum
    ## The following code calculates a fovy based on a right triangle
    ## where the vertical leg is length 1 (on the frustum) and the
    ## other is eyedist (from the frustum).  The fovy is double that
    ## angle, converted to degrees.
    fovy = 2*math.atan(1/eyedist)*180/M_PI
    print "eyez is %f, eye distance is %f so fovy = %f degrees" % (
        RightWinEyeZ,eyedist,fovy)
    glMatrixMode(GL_PROJECTION);
    glLoadIdentity();
    gluPerspective(fovy,ar,eyedist,eyedist+2);
    glMatrixMode(GL_MODELVIEW);
    glLoadIdentity();
    gluLookAt(0,0,RightWinEyeZ, 0,0,0, 0,1,0);

    twColorName(TW_MAGENTA);
    glutWireCube(1);
  
    glFlush();
    glutSwapBuffers();

### These hold window identifiers, which are integers returned by the
### windowing system.
parentWin = 0
leftWin = 0
rightWin = 0

def redisplayAll():
    glutSetWindow(leftWin)      # makes leftWin be the current window
    glutPostRedisplay();        # say it needs redisplay
    glutSetWindow(rightWin);
    glutPostRedisplay();

def modifyM(key, x, y):
    '''Modify the distance that the eye is from the cube in the right window.'''
    global AdditionalEyeDistance
    if key == '+':
        AdditionalEyeDistance += 1.0
    elif key == '-':
        AdditionalEyeDistance -= 1.0
        if AdditionalEyeDistance < 0.0:
            AdditionalEyeDistance = 0.0
    print "AdditionalEyeDistance is now ", AdditionalEyeDistance
    eyedist = LeftWinEyeZ+AdditionalEyeDistance
    #print "Relative size of front and back is ", 2*(M-2)/(M-1)
    #print " Back projects to size ", 2*(M-2)/(M+1)
    redisplayAll();

def main():
    global parentWin, leftWin, rightWin
    glutInit(sys.argv)
    glutInitDisplayMode( GLUT_DOUBLE | GLUT_RGB | GLUT_DEPTH)
    # create a main window with enough room for both subwindows and the gaps
    glutInitWindowSize(
      GAP+leftWinWidth+GAP+rightWinWidth+GAP, # widths from left to right
      GAP+leftWinHeight+GAP                   # heights from top to bottom
      )
    parentWin = glutCreateWindow(sys.argv[0]);
    glutDisplayFunc(parentDisplay);

    twBoundingBox(-1,1,-1,1,-1,1); # twMainInit will complain if BB not set
    twMainInit();

    ## glutCreateSubWindow returns the ID of the sub-window and takes 
    ## arguments:(int parent,int x,int y,int width,int height).
    ## Origin at x,y; height equals distance from y _down_
    leftWin = glutCreateSubWindow(parentWin,GAP,GAP,leftWinWidth,leftWinHeight);

    ## Each time you create a new window you need to give it its own 
    ## callbacks and initialization. These can be the same as those
    ## from other windows but you must call them each time
    twBoundingBox(-10,10,-10,10,-10,10);
    glutDisplayFunc(leftDisplay);
    twMainInit(); 

    rightWin = glutCreateSubWindow(parentWin,GAP*2+leftWinWidth,GAP,rightWinWidth,rightWinHeight);
    glutDisplayFunc(rightDisplay);
    twMainInit();
    twKeyCallback('+',modifyM,'Increase distance in right window')
    twKeyCallback('-',modifyM,'Decrease distance in right window')
    glLineWidth(2);
  
    redisplayAll(); 
    glutMainLoop();

if __name__ == '__main__':
    main()
